File: crates/simcity_core/src/game/map/coords.rs
//! Canonical tile <-> world mapping. THE ONLY place this formula lives.
//!
//! The sim's logical world is the XY plane; the map is centered on the world
//! origin. Every producer of world positions (render sync, spawns, overlays,
//! persistence, picking) must go through these functions so the projection
//! can change in exactly one place (pseudo-3D migration, phase 2+).

use bevy::math::{Vec2, Vec3};
use bevy::prelude::{Camera, GlobalTransform};

use super::{MapConfig, TilePos};

/// World-space center of tile (0,0). The map is centered on the world origin.
pub fn map_origin(cfg: &MapConfig) -> Vec2 {
    Vec2::new(
        -((cfg.width - 1) as f32) * cfg.tile_size * 0.5,
        -((cfg.height - 1) as f32) * cfg.tile_size * 0.5,
    )
}

/// Center of `tile` in logical world coordinates.
pub fn tile_to_world(cfg: &MapConfig, tile: TilePos) -> Vec2 {
    tile_f_to_world(cfg, tile.x as f32, tile.y as f32)
}

/// Fractional tile coordinates -> logical world (multi-tile footprint centers,
/// sub-tile offsets like lane centers and parking bays).
pub fn tile_f_to_world(cfg: &MapConfig, tile_x: f32, tile_y: f32) -> Vec2 {
    map_origin(cfg) + Vec2::new(tile_x * cfg.tile_size, tile_y * cfg.tile_size)
}

/// Ray parameter `t` where `origin + t*dir` crosses the ground plane (z = 0).
/// None if the ray is parallel to the ground or points away from it.
pub fn ray_ground_t(origin: Vec3, dir: Vec3) -> Option<f32> {
    if dir.z.abs() < 1e-6 {
        return None;
    }
    let t = -origin.z / dir.z;
    (t >= 0.0).then_some(t)
}

/// Viewport point -> logical ground plane (the sim's XY world at z = 0).
/// The pseudo-3D counterpart of `viewport_to_world_2d`: valid for any camera.
pub fn viewport_to_ground(
    camera: &Camera,
    cam_gt: &GlobalTransform,
    viewport: Vec2,
) -> Option<Vec2> {
    let ray = camera.viewport_to_world(cam_gt, viewport).ok()?;
    let t = ray_ground_t(ray.origin, *ray.direction)?;
    Some((ray.origin + *ray.direction * t).truncate())
}

/// Inverse mapping with round-to-nearest semantics (picking). None outside the map.
pub fn world_to_tile(cfg: &MapConfig, world: Vec2) -> Option<TilePos> {
    let local = world - map_origin(cfg);
    let x = (local.x / cfg.tile_size).round() as i32;
    let y = (local.y / cfg.tile_size).round() as i32;
    if x < 0 || y < 0 || x >= cfg.width || y >= cfg.height {
        return None;
    }
    Some(TilePos { x, y })
}

File: packages/sim/src/map/coords.ts
// Canonical tile <-> world mapping: port of simcity_core/src/game/map/coords.rs. World coordinates
// are f32 in Rust, so every operation is rounded to f32; the map is centred on the world origin.
import type { TilePos } from '../commands';

export interface MapConfig {
  readonly width: number;
  readonly height: number;
  /** `f32`. */
  readonly tileSize: number;
}

export const DEFAULT_MAP_CONFIG: MapConfig = { width: 128, height: 128, tileSize: 16 };

export interface Vec2 {
  readonly x: number;
  readonly y: number;
}

const f32 = Math.fround;

/** `f32::round`: half away from zero. Exact for f32 inputs: `v + 0.5` needs no more bits than a double has. */
function roundHalfAwayFromZero(v: number): number {
  return v < 0 ? -Math.floor(-v + 0.5) : Math.floor(v + 0.5);
}

/** World-space centre of tile (0,0). */
export function mapOrigin(cfg: MapConfig): Vec2 {
  const ts = f32(cfg.tileSize);
  return {
    x: f32(f32(-(cfg.width - 1) * ts) * 0.5),
    y: f32(f32(-(cfg.height - 1) * ts) * 0.5),
  };
}

/** Fractional tile coordinates -> world (footprint centres, sub-tile offsets). */
export function tileFToWorld(cfg: MapConfig, tileX: number, tileY: number): Vec2 {
  const origin = mapOrigin(cfg);
  const ts = f32(cfg.tileSize);
  return {
    x: f32(origin.x + f32(f32(tileX) * ts)),
    y: f32(origin.y + f32(f32(tileY) * ts)),
  };
}

/** Centre of `tile` in world coordinates. */
export function tileToWorld(cfg: MapConfig, tile: TilePos): Vec2 {
  return tileFToWorld(cfg, tile.x, tile.y);
}

/** Inverse mapping with round-to-nearest semantics (picking); `undefined` outside the map. */
export function worldToTile(cfg: MapConfig, world: Vec2): TilePos | undefined {
  const origin = mapOrigin(cfg);
  const ts = f32(cfg.tileSize);
  const x = roundHalfAwayFromZero(f32(f32(world.x - origin.x) / ts)) | 0;
  const y = roundHalfAwayFromZero(f32(f32(world.y - origin.y) / ts)) | 0;
  if (x < 0 || y < 0 || x >= cfg.width || y >= cfg.height) return undefined;
  return { x, y };
}
